### PR TITLE
lavamoat-protect WIP round 1

### PR DESCRIPTION
**Pull request by legobeat**
@naugtur

### CODE REVIEW

File: packages/allow-scripts/src/setup.js [hunk 1]
```diff
@@ -68,23 +67,54 @@ function writeRcFile (pkgManagerDefault = 'npm') {
...
+  for (const [k, v] of Object.entries(input)) {
+    oldConf[k] = typeof v === 'object' && typeof oldConf[k] === 'object'
+      ? { ...oldConf[k], ...v }
+      : v
   }
+  fs.writeFileSyn(p, JSON.stringify(mergedConf, undefined, 2), { encoding: 'utf-8' })
+}
 
-  let result = spawnSync(cmd, cmdArgs, {})
+function patchPackageJson (runPkgManager) {
```
Comment (legobeat) [inline packages/allow-scripts/src/setup.js]: Naming overload with `updatePackageJson`/`patchPackageJson`. TBD if they will be merged or either renamed.

File: packages/allow-scripts/src/setup.js [hunk 2]
```diff
@@ -68,23 +67,54 @@ function writeRcFile (pkgManagerDefault = 'npm') {
...
-    console.log('@lavamoat/allow-scripts:: Added dependency @lavamoat/preinstall-always-fail.')
+    switch (runPkgManager) {
+      case 'npm':
+        cmd = 'npm'
+        cmdArgs = []
+        break
+      case 'yarn1':
+      case 'yarn3':
+        cmd = 'yarn'
+        cmdArgs = []
```
Comment (legobeat) [inline packages/allow-scripts/src/setup.js]: I considered detecting presence of `setup` and running that instead but best not assuming it has lavamoat semantics if present (ducktyping.jpg)

File: packages/lavamoat-protect/lib/setup-scripts.js [hunk 3]
```diff
@@ -1,40 +1,24 @@
...
     console.warn('Existing script `lavamoat-postinstall` detected. This is indicative of inconsistently configured LavaMoat')
-    
+
     if(!(await promptly.confirm('Would you like to override the existing script?'))) {
       throw new Error('User exit');
     }
   }
 
-  pkgConf.scripts['lavamoat-postinstall'] = './node_modules/@lavamoat/allow-scripts/src/cli.js'
+  patchPackageJson(pkgManager)
```
Comment (naugtur) [inline packages/lavamoat-protect/lib/setup-scripts.js]: if patchPackageJson can run pkg commands, we'd need it to be after the RC file is crated.

File: packages/allow-scripts/src/cli.js [hunk 4]
```diff
@@ -32,7 +32,7 @@ async function start () {
     }
     case 'setup': {
       writeRcFile()
-      addPreinstallAFDependency()
+      patchPackageJson(existsSync('./.npmrc') ? 'npm' : 'yarn1')
```
Comment (naugtur) [inline packages/allow-scripts/src/cli.js]: we have a much better implementation of this and if we're in need of it here, we should use that.
I thought we'd eliminate the install step here and let the user run whatever installation they like.
(note there's a lot of options. `npm install` or `npm ci`, `yarn` alone or with options to strictly adhere to lockfile etc.)

Comment (legobeat) [reply packages/allow-scripts/src/cli.js]: Ah, the main idea here was to retain previous behavior of the allow-scripts cli - either way, this logic is now removed.

File: packages/lavamoat-protect/lib/setup-scripts.js (hunk 3 above)
Comment (legobeat) [reply packages/lavamoat-protect/lib/setup-scripts.js]: Updated

**Review by naugtur — APPROVED** [review]
LGTM

**Pull request merged**